File: algorithms/java/src/test/java/leetcode/LC111MinimumDepthOfBinaryTree.java
package leetcode.lc111_minimum_depth_of_binary_tree;

import java.util.ArrayList;
import java.util.List;

import org.junit.jupiter.api.Test;


import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/minimum-depth-of-binary-tree/
 * #easy
 */
public final class LC111MinimumDepthOfBinaryTree {
    public final class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(final int x) {
            val = x;
        }
    }

    public final class Solution {
        private class AnnotatedNode {
            public TreeNode node;
            public int depth;

            public AnnotatedNode(final TreeNode node, final int depth) {
                this.node = node;
                this.depth = depth;
            }
        }

        public int minDepth(final TreeNode root) {
            if (root == null) {
                return 0;
            }
            List<AnnotatedNode> queue = new ArrayList<>();
            queue.add(new AnnotatedNode(root, 1));
            while (queue.size() > 0) {
                AnnotatedNode annotated = queue.remove(0);
                if (annotated.node.left == null && annotated.node.right == null) {
                    return annotated.depth;
                } else {
                    if (annotated.node.left != null) {
                        queue.add(new AnnotatedNode(annotated.node.left, annotated.depth + 1));
                    }
                    if (annotated.node.right != null) {
                        queue.add(new AnnotatedNode(annotated.node.right, annotated.depth + 1));
                    }
                }
            }
            return -1;
        }
    }

    @Test
    public void test_example() throws Exception {
        TreeNode n3 = new TreeNode(3);
        TreeNode n7 = new TreeNode(7);
        TreeNode n9 = new TreeNode(9);
        TreeNode n15 = new TreeNode(15);
        TreeNode n20 = new TreeNode(20);
        n3.left = n9;
        n3.right = n20;
        n20.left = n15;
        n20.right = n7;
        n7.left = n7.right = n9.left = n9.right = n15.left = n15.right = null;
        assertEquals(2, new Solution().minDepth(n3));
    }
}

File: algorithms/code/leetcode/lc111_minimum_depth_of_binary_tree/lc111_minimum_depth_of_binary_tree.rs
// https://leetcode.com/problems/binary-tree-paths/
// #easy

use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<Self>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }
}

struct AnnotatedNode {
    node: Rc<RefCell<TreeNode>>,
    depth: i32,
}

pub struct Solution;

impl Solution {
    pub fn min_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        if root.is_none() {
            return 0;
        }
        let mut queue = VecDeque::new();
        queue.push_back(AnnotatedNode {
            node: root.unwrap(),
            depth: 1,
        });
        let mut depth: i32 = 0;
        while let Some(annotated) = queue.pop_front() {
            let annotated_node = annotated.node.borrow();
            let left = annotated_node.left.clone();
            let right = annotated_node.right.clone();
            if left.is_none() && right.is_none() {
                depth = annotated.depth;
                break;
            }
            if let Some(left_node) = left {
                queue.push_back(AnnotatedNode {
                    node: left_node,
                    depth: annotated.depth + 1,
                });
            }
            if let Some(right_node) = right {
                queue.push_back(AnnotatedNode {
                    node: right_node,
                    depth: annotated.depth + 1,
                });
            }
        }
        depth
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        let n15 = TreeNode::new(15, None, None);
        let n7 = TreeNode::new(7, None, None);
        let n9 = TreeNode::new(9, None, None);
        let n20 = TreeNode::new(20, n15, n7);
        let n3 = TreeNode::new(3, n9, n20);
        assert_eq!(2, Solution::min_depth(n3));
    }

    #[test]
    fn test_left_nothing() {
        let right = TreeNode::new(7, None, None);
        let root = TreeNode::new(3, None, right);
        assert_eq!(2, Solution::min_depth(root));
    }

    #[test]
    fn test_right_nothing() {
        let left = TreeNode::new(7, None, None);
        let root = TreeNode::new(3, left, None);
        assert_eq!(2, Solution::min_depth(root));
    }

    #[test]
    fn test_nothing() {
        assert_eq!(0, Solution::min_depth(None));
    }
}
